enhancements/microshift/microshift-apiserver-certs.md

```diff
@@ -0,0 +1,326 @@
+SAN). Because of how the certificate lookup works for incoming connections,
+there can only be one certificate associated to an IP address. Because of
+MicroShift network architecture, external and internal clients appear to use
+the same server IP. Connections from both types of clients would use the same
+certificate to authenticate the connection.
+
+### User Stories
+* As a MicroShift administrator/user, I want authentication by using the
+hostname.
+* As a MicroShift administrator/user, I want authentication by a name
+registered in a DNS.
+* As a MicroShift administrator/user, I want authentication by its IP address.
+* As a MicroShift application developer, I want authentication by the service
+name, `kubernetes`.
+* As a MicroShift application developer, I want authentication by the service
+IP.
+
+### Goals
+* Allow authentication from all possible sources to API server.
+
+### Non-Goals
+N/A
+
+## Proposal
+MicroShift uses several serving certificates for internal and external
+communication. Among them we can find:
+* Kubelet
+* Route controller manager
+* Router default
+* Etcd
+* API server external
+* API server localhost
+* API server service network
+
+Each of these certificates is signed by a different CA. The ones we are
+focusing on are the API server serving certificates.
+
+Proposal includes merging all API server certificates into a single one,
+allowing valid authentication regardless of the client: external or internal.
+This enables the use of IPs in addition to hostnames, which is part of
+MicroShift requirements.
+
+### Current certificate strategy
+#### How API server builds certificates configuration
+The current strategy matches the one from OpenShift Container Platform (OCP),
+where there are different trust domains depending on who wants to connect to
+API server. Each client will get their corresponding certificate based on
+the trust domain in which they are.
+
+A description of how the API server configures certificates follows.
+
+When creating the API server instance there is a [default certificate](https://github.com/openshift/microshift/blob/main/pkg/controllers/kube-apiserver.go#L136)
+which is the serving service network. The rest of the certificates are
+configured [here](https://github.com/openshift/microshift/blob/main/pkg/controllers/kube-apiserver.go#L184-L202)
+along with the TLS versions and cipher suites. These go into an upstream
+library called `dynamiccertificates`. If certificates have names on them (which
+they do), the package [creates](https://github.com/kubernetes/apiserver/blob/master/pkg/server/dynamiccertificates/tlsconfig.go#L209)
+a 1:1 mapping of names to certificates. There is one entry per name in each
+certificate, including Common Name (CN) and Subject Alternative Names (SAN).
+Note there can only be one certificate per name, as otherwise there would be
+ambiguity.
+
+#### How API server selects certificates for incoming connections
```

Comment:
the kube-apiserver also contains custom clientHello logic that allows binding IPs via SNI-like configuration from hostnames.

Reply:
I'm not sure what that means. How does that apply to the problem?

Reply:
> I'm not sure what that means. How does that apply to the problem?

It means this summary lacks nuance.  While normally true of SNI, the kube-apiserver allows SNI configuration based on IPs, which means that if SNI is consistently configured, it is possible to predict what the resulting ca.crt needs to be inside of pods without resorting to "big ball of trust" and a single serving certificate

Reply:
OK, the team hasn't found how to configure that. Can you help with references?

Reply:
@deads2k Do you mean the `tls-sni-cert-key` option? I also tried but came accross the same issue. The [NamedCertificates](https://github.com/openshift/microshift/blob/main/pkg/controllers/kube-apiserver.go#L184) approach seems to yield the same result in the TLS config.
External connections try to use the node IP to connect to API server, and internal ones do so too (because of the endpoint slice having the node IP in the address list), so how to select the certificate for different connections that appear to differ only by the source, not the destination?
Or is there another way I dont know of?